front_end/sources/JavaScriptSourceFrame.js

```diff
@@ -1528,7 +1528,23 @@ Sources.JavaScriptSourceFrame = class extends Sources.UISourceCodeFrame {
       --linesToCheck;
       const locations = await this._breakpointManager.possibleBreakpoints(
           this._debuggerSourceCode, new TextUtils.TextRange(lineNumber, 0, lineNumber, lineLength));
+      // DDC splits variable declarations with initialization into a
+      // declaration and initialization in the constructor. So the
```

Comment:
I'm not totally following this comment, what does the Dart code look like? Should DDC do something else?

I think this comment is about instance fields on a class (rather than "variable declarations")? Assuming that's correct (if not, disregard the rest of this comment :) )


Fields are weird for a few reasons. They don't map 1:1 -- field initializers appear in every constructor. So DDC can either choose arbitrarily to only map one of them, or else it maps all of those JS locations back to the same Dart line.

The initializer expression is mapped, same for the field name. The name is for hover support -- if we fix how hover is implemented, we can remove the hover markings.

But both of those should things should be mapped to the same JS line. For example:

```dart
class C {
  final x = 123;
  final y;
  C() : y = 456;
  C.y(this.y);
}
```
in the JS code for `C()` it should have these mappings:
```js
this.x = 123; // `this.x` maps to field name `x` in Dart
              // `123` maps to the field initializer expression
this.y = 456; // `y` maps to the `y` in constructor initializer `: y = 456`
              // `456` maps to the initializer expression
```
Now the catch is that Dart line `final x = 123` will have pieces of it mapped into both `C()` and `C.y()` in this example. Maybe that's the problem?

Reply:
Oh gotcha! thanks that helps a lot, makes total sense. Weird that Chrome implemented things that way, but I like your workaround.

Yeah I forgot about the virtual getters/setters, good catch. :)

Originally I didn't map the virtual getters/setters in DDC, since it doesn't correspond to Dart source. Someone added that to DDK. I think the idea was: so that you could "step in" and "step out" of the get/set. If we don't map those virtual accessors, when you step in (legal in JS since there is a getter) Chrome will look for a line mapping, not find it, and then grab some nearby line (like the start of the next Dart method or the end of the Dart class).

So while we could remove those getter/setter mappings, I think probably after your fix we don't need to? So we can kinda get the best of both worlds (stepping & breakpoints will work reasonably)

Reply:
I think we should be able to leave them. It does do the odd thing that sometimes it leaves a whole bunch of disabled breakpoints over that range, but it doesn't actually break anything. You can still set normal breakpoints on those lines, etc.